pkg/controller/kubedirectorcluster/cluster_restore.go

```diff
@@ -0,0 +1,411 @@
+) {
+
+	// We need to check for the cluster (headless) service, the statefulsets,
+	// and the per-member services associated with statefulset pods.
+	// Everything else will be created by the statefulset controller.
+
+	resourcesPresent := clusterServiceExists(reqLogger, cr)
+	resourcesPresent = resourcesPresent && roleResourcesExist(reqLogger, cr)
+
+	if resourcesPresent {
+		if cr.Status.RestoreProgress.AwaitingResources {
+			cr.Status.RestoreProgress.AwaitingResources = false
+			shared.LogInfo(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: component resources are present",
+			)
+		}
+	} else {
+		cr.Status.RestoreProgress.AwaitingResources = true
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"being restored: awaiting component resources",
+		)
+	}
+}
+
+// clusterServiceExists looks to see if a headless service named in the
+// status exists.
+func clusterServiceExists(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	clusterService := cr.Status.ClusterService
+	if clusterService != "" {
+		_, serviceErr := observer.GetService(
+			cr.Namespace,
+			clusterService,
+		)
+		if serviceErr != nil {
+			shared.LogInfof(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: cluster service %s does not exist",
+				clusterService,
+			)
+			return false
+		}
+	}
+	return true
+}
+
+// roleResourcesExist looks to see if a statefulsets named in the
+// status exist, along with the necessary per-member services.
+func roleResourcesExist(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	for _, roleStatus := range cr.Status.Roles {
+		if roleStatus.StatefulSet != "" {
+			_, statefulSetErr := observer.GetStatefulSet(
+				cr.Namespace,
+				roleStatus.StatefulSet,
+			)
+			if statefulSetErr != nil {
+				shared.LogInfof(
+					reqLogger,
+					cr,
+					shared.EventReasonCluster,
+					"being restored: statefulset %s does not exist",
+					roleStatus.StatefulSet,
+				)
+				return false
+			}
+		}
+		for _, memberStatus := range roleStatus.Members {
+			memberService := memberStatus.Service
+			if memberService != "" && memberService != zeroPortsService {
+				_, serviceErr := observer.GetService(
+					cr.Namespace,
+					memberService,
+				)
+				if serviceErr != nil {
+					shared.LogInfof(
+						reqLogger,
+						cr,
+						shared.EventReasonCluster,
+						"being restored: member service %s does not exist",
+						memberService,
+					)
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+// checkRestoreDone will try to clear the being-restored label if all the
+// "waiting" flags are false. If that fails it will populate the error field
+// of the restore progress object. Return true if and only if we have
+// successfully cleared that label.
+func checkRestoreDone(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	if cr.Status.RestoreProgress.AwaitingApp ||
```

Comment:
can we come here after marking error in last run? should we check error also here?

Reply:
I think even if there was previously an error, it's ok to let KD try again (on next handler) to auto-resume reconciliation. It will basically keep trying until it succeeds, or some of the dependencies go back to false again, or the kdcluster is deleted.

There are probably only two broad scenarios where it might be likely to fail an attempt and then succeed on a subsequent attempt:

* Temporary failure to interact with the API server for whatever reason.

* Validation failure that can be humanly resolved. It's tough to imagine many cases for this one, especially if you're just doing "back up everything" and then "restore everything" cases for a namespace. However if you were trying to do something more fiddly, like only back up kdclusters and manually manage the presence/absence of kdapps, then you might get in a situation where the kdapp with the name referenced by the kdcluster is different from the one that was being used at backup time. In that (contrived) case, validation will fail until you get the correct kdapp back in there.

Reply:
ok makes sense. We are okay here then